subxt/src/rpc/rpc_client_t.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::error::RpcError;
+use futures::Stream;
+use serde_json::value::RawValue;
+use std::{
+    future::Future,
+    pin::Pin,
+};
+
+/// Any RPC client which implements this can be used in our [`super::Rpc`] type
+/// to talk to a node.
+///
+/// This is a low level interface which expects an already-serialized set of params,
+/// and returns an owned but still-serialized [`RawValue`], deferring deserialization to
+/// the caller. This is the case because we want the methods to be object-safe (which prohibits
+/// generics), and want to avoid any unnecessary allocations in serializing/deserializing
+/// parameters.
+pub trait RpcClientT: Send + Sync + 'static {
+    /// Make a raw request for which we expect a single response back from. The params will
+    /// be provided in the form of a pre-encoded JSON array.
+    fn request_raw<'a>(
+        &'a self,
+        method: &'a str,
+        params: Box<RawValue>,
```

Comment:
nit: Can we also receive no parameters here?

Reply:
I think the intention is that you are supposed to use the `rpc_params` macro that James wrote which takes care of that for you.

However, I agree that creating a `RawValue` from the an empty `JSON String` is quite awkward because you need to allocate a String for it.

Reply:
I have just tweaked these things to accept `Option<Box<RawValue>>`, to allow subxt to avoid the unnecessary allocations when there are no params :)